Implemented duplicate patients management.

sunsetsystems opened the pull request:
Identification and correction of duplicate patients involves the following:

 1. Column "dupscore" in the patient_data table was created to contain a value representing
    the likelihood that the patient duplicates any of the patients created earlier (that is,
    with a lesser "pid" value). It may have any of the following values:

    o -9, initially assigned, indicating that it needs to be computed.
    o -1, indicating that the patient has been manually flagged as not duplicating a previous patient.
    o A score in the range 0-40 computed by the SQL expression found in library/dupscore.inc.php.
      A value of 12 or higher is considered to be very likely a duplicate.

 2. The script library/dupscore.inc.php which provides the SQL expression needed for scoring.

 3. The command-line tool contrib/util/dupscore.cli.php which should be run to initially
    assign all patient_data.dupscore values, and may be run from time to time to update
    or recompute them. See the source of that script for its usage syntax.

 4. The web tool Administration -> Patient -> Manage Duplicates.
    For each patient with a dupscore of 7 or higher, it shows below it all the corresponding
    matching patients. Thus you see a collection of patient groups.
    Within each group you can choose to mark the first patient as not a duplicate, or you can
    merge it with one of the other patients. You do that by choosing "Merge and Keep" or
    "Merge and Discard" for the other patient, which brings up the Merge Patients page to
    complete the merge. That page describes in more detail what it will do.

<!--
(Thanks for sending a pull request!
-->

<!-- Add the issue number that is fixed by this PR (In the form Fixes #123) -->
Fixes #

#### Short description of what this resolves:


#### Changes proposed in this pull request:

sunsetsystems commented:
Not done testing this yet, but wanted to go ahead and get it up here for comments.

sunsetsystems commented on library/patient.inc:
```diff
@@ -8,9 +8,10 @@
+ * @author    Rod Roark <rod@sunsetsystems.com>
  * @copyright Copyright (c) 2018-2019 Brady Miller <brady.g.miller@gmail.com>
  * @copyright Copyright (c) 2019 Sherwin Gaddis <sherwingaddis@gmail.com>
- * @copyright Copyright (c) 2018-2021 Stephen Waite <stephen.waite@cmsvt.com>
```
Oops sorry Stephen, I'll put it back.

mdsupport commented:
1. If the fields considered to flag duplicates come from user configurable list, that will make the approach more flexible.
2. All externally assigned numbers - ssn, driver's license, insurance policy are good flags. May be insurance information can be pulled in as well.
3. Most fields are strings. So use of soundex will produce simpler and better results. 
4. Although it makes for slow response, use of IN function simplifies SQL - e.g. ph IN (ph1, ph2, ph3, ph4).
5. If you must use equality operator, be sure to eliminate case variation. Have never figured out why some people insist on using all uppercase letters for data entry!
6. Another good indicator are the addresses. We used api's to get GIS coordinates. Maybe others have better idea.

sunsetsystems commented:
@mdsupport Great suggestions! I'll ponder these a bit, though some of it may have to wait for future improvements.

sunsetsystems commented:
Alright I changed to soundex in a couple of places, and simplified the phone number tests as you suggested.

The fields of interest do seem to have case-insensitive collation so that should be OK.

I like the idea of a user configurable list but let's save that as a future improvement. I wonder if the corresponding SQL fragments should be in the lists also, or if that's just asking for trouble. :)

bradymiller commented on contrib/util/dupscore.cli.php:
```diff
@@ -0,0 +1,155 @@
+ * @copyright Copyright (c) 2021 Rod Roark <rod@sunsetsystems.com>
+ * @license   https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3
+ */
+
```
for command line php script recommend placing exit at top (which can be uncommented when used) and ensuring only can be run via commandline. see here for example:
https://github.com/openemr/openemr/blob/master/contrib/util/ccda_import/import_ccda.php#L39-L42

sunsetsystems replied:
@bradymiller Would the test for php_sapi_name() be sufficient? That will ensure it's not being run via the web server. Since it only computes dupscore values it won't do a lot of damage if run accidentally from the command line.

bradymiller replied:
@sunsetsystems , sounds good. Agreed very benign script at this point.

bradymiller replied:
btw, this type of thing could also be run as a background service in OpenEMR (would push us to then make a mechanism in the background service for times of day when ok to run since wouldn't want to run this during the day).

sunsetsystems replied:
Yes I thought about the background service possibility but as you say it would need some work. The important thing is to run it after upgrading the site. New patients will have their dupscores computed at add time but those may become outdated if demographics of previous patients are fixed, which is why the script should run from time to time.

bradymiller commented on contrib/util/dupscore.cli.php:
```diff
@@ -0,0 +1,155 @@
+if (!is_file($confname)) {
+    die("File not found: " . $confname . "\n");
+}
+
```
could consider bringing in globals.php context with auth ignored. for example:
https://github.com/openemr/openemr/blob/master/contrib/util/ccda_import/import_ccda.php#L67-L68
(then no need to deal with making sql function in addition to sql injection support, multisites easier to support, and can expand functionality of this script will be easier)

sunsetsystems replied:
Wow I didn't know I could do that. Will do. Also the xl() function will be available.

bradymiller commented on contrib/util/dupscore.cli.php:
```diff
@@ -0,0 +1,155 @@
+
+    mysqli_free_result($res1);
+    foreach ($scores as $pid => $score) {
+        sqlExec($link, "UPDATE patient_data SET dupscore = $score WHERE pid = $pid");
```
would pass the variable through the sql escaping function with quotes around them (mainly to ensure any expansion of functionality in this script will then follow this)

sunsetsystems replied:
Will use binding.

bradymiller commented on interface/main/tabs/menu/menus/standard.json:
```diff
@@ -852,7 +852,19 @@
+            "requirement": 0,
+            "acl_req": [
+              "admin",
+              "delete"
```
is this a new aco?

sunsetsystems replied:
Hmm it will be. :) I'll change it to super for now.

bradymiller commented on interface/patient_file/manage_dup_patients.php:
```diff
@@ -0,0 +1,304 @@
+    ?>
+ <tr bgcolor='<?php echo $bgcolor; ?>'>
+  <td class="detail" bgcolor="#dddddd">
+   <select onchange='selchange(this, <?php echo "$pid, " . $row['pid']; ?>)' style='width:100%'>
```
```
<select onchange='selchange(this, <?php echo attr_js($pid); ?>,<?php echo attr_js($row['pid']); ?>)' style='width:100%'>
```

bradymiller commented on interface/patient_file/manage_dup_patients.php:
```diff
@@ -0,0 +1,304 @@
+  <td class="detail" align="right">
+    <?php echo text($myscore); ?>
+  </td>
+  <td class="detail" align="right" onclick="openNewTopWindow(<?php echo $row['pid']; ?>)"
```
```
<td class="detail" align="right" onclick="openNewTopWindow(<?php echo attr_js($row['pid']); ?>)"
```

bradymiller commented on interface/patient_file/manage_dup_patients.php:
```diff
@@ -0,0 +1,304 @@
+    <?php echo text($myscore); ?>
+  </td>
+  <td class="detail" align="right" onclick="openNewTopWindow(<?php echo $row['pid']; ?>)"
+    title="Click to open in a new window or tab" style="color:blue;cursor:pointer">
```
translate

bradymiller commented on interface/patient_file/manage_dup_patients.php:
```diff
@@ -0,0 +1,304 @@
+    <?php echo text($ptname); ?>
+  </td>
+  <td class="detail">
+    <?php echo oeFormatShortDate($row['DOB']); ?>
```
text()

bradymiller commented on interface/patient_file/manage_dup_patients.php:
```diff
@@ -0,0 +1,304 @@
+    <?php echo text($phones); ?>
+  </td>
+  <td class="detail">
+    <?php echo oeFormatShortDate($row['regdate']); ?>
```
text()

bradymiller commented on interface/patient_file/manage_dup_patients.php:
```diff
@@ -0,0 +1,304 @@
+}
+
+if (!AclMain::aclCheckCore('admin', 'delete')) {
+    die(xl("Unauthorized access."));
```
xlt()

bradymiller commented on interface/patient_file/manage_dup_patients.php:
```diff
@@ -0,0 +1,304 @@
+    }
+}
+
+if (!AclMain::aclCheckCore('admin', 'delete')) {
```
new aco?

bradymiller commented on interface/patient_file/manage_dup_patients.php:
```diff
@@ -0,0 +1,304 @@
+
+<script type="text/javascript" src="../../library/textformat.js?v=<?php echo $v_js_includes; ?>"></script>
+<script type="text/javascript" src="../../library/dialog.js?v=<?php echo $v_js_includes; ?>"></script>
+<script type="text/javascript" src="../../library/js/jquery-1.9.1.min.js"></script>
```
these are already included in the setupHeader call, so should remove these

bradymiller commented on interface/patient_file/manage_dup_patients.php:
```diff
@@ -0,0 +1,304 @@
+    }
+});
+
+var mypcc = '<?php echo $GLOBALS['phone_country_code'] ?>';
```
remove this if can, if not then change to (note quotes were removed):
```
var mypcc = <?php echo js_escape($GLOBALS['phone_country_code']); ?>;
```

bradymiller commented on interface/patient_file/manage_dup_patients.php:
```diff
@@ -0,0 +1,304 @@
+  if (sel.value == '') return;
+  top.restoreSession();
+  if (sel.value == 'MK') {
+    window.location = 'merge_patients.php?pid1=' + rowpid + '&pid2=' + toppid;
```
```
window.location = 'merge_patients.php?pid1=' + encodeURIComponent(rowpid) + '&pid2=' + encodeURIComponent(toppid);
```

bradymiller commented on interface/patient_file/manage_dup_patients.php:
```diff
@@ -0,0 +1,304 @@
+    window.location = 'merge_patients.php?pid1=' + rowpid + '&pid2=' + toppid;
+  }
+  else if (sel.value == 'MD') {
+    window.location = 'merge_patients.php?pid1=' + toppid + '&pid2=' + rowpid;
```
incorporate encodeURIComponent() in above line

bradymiller commented on interface/patient_file/manage_dup_patients.php:
```diff
@@ -0,0 +1,304 @@
+
+<!-- form used to open a new top level window when a patient row is clicked -->
+<form name='fnew' method='post' target='_blank'
+ action='../main/main_screen.php?auth=login&site=<?php echo attr($_SESSION['site_id']); ?>'>
```
```
action='../main/main_screen.php?auth=login&site=<?php echo attr_url($_SESSION['site_id']); ?>'>
```

bradymiller commented on interface/patient_file/merge_patients.php:
```diff
@@ -8,23 +8,27 @@
 $PRODUCTION = true;
 
-if (!AclMain::aclCheckCore('admin', 'super')) {
+if (!AclMain::aclCheckCore('admin', 'delete')) {
```
new aco?

bradymiller commented on interface/patient_file/merge_patients.php:
```diff
@@ -269,23 +273,56 @@ function mergeRows($tblname, $colname, $source_pid, $target_pid)
 </p>
 
-<form method='post' action='merge_patients.php'>
+<form method='post' action='merge_patients.php?<?php echo "pid1=$form_pid1&pid2=$form_pid2"; ?>'>
```
```
<form method='post' action='merge_patients.php?<?php echo "pid1=" . attr_url($form_pid1) . "&pid2=" . attr_url($form_pid2); ?>'>
```

bradymiller commented on interface/patient_file/merge_patients.php:
```diff
@@ -294,8 +331,11 @@ function mergeRows($tblname, $colname, $source_pid, $target_pid)
+    value=' (<?php echo attr($target_string); ?>)'
+    onclick='sel_patient(this, this.form.form_target_pid)'
+    title='<?php echo xla('Click to select patient'); ?>' readonly />
+   <input type='hidden' name='form_target_pid' value='<?php echo $target_pid; ?>' />
```
attr()

bradymiller commented on interface/patient_file/merge_patients.php:
```diff
@@ -307,10 +347,10 @@ function mergeRows($tblname, $colname, $source_pid, $target_pid)
     onclick='sel_patient(this, this.form.form_source_pid)'
     title='<?php echo xla('Click to select patient'); ?>' readonly />
-   <input type='hidden' name='form_source_pid' value='0' />
+   <input type='hidden' name='form_source_pid' value='<?php echo $source_pid; ?>' />
```
attr()

bradymiller commented:
review complete :)

sunsetsystems commented:
@bradymiller Thanks for the review. I think your issues are all addressed now.

bradymiller commented on contrib/util/dupscore.cli.php:
```diff
@@ -0,0 +1,123 @@
+}
+
+// Bring in the getDupScoreSQL() function.
+require_once($args['webdir'] . "/library/dupscore.inc.php");
```
might as well bring this in after globals.php (ie. standard openemr script organization and then can use a openemr context variable for the path)

sunsetsystems replied:
Done.

bradymiller commented on contrib/util/dupscore.cli.php:
```diff
@@ -0,0 +1,123 @@
+    $query1 = "SELECT p1.pid, MAX(" . getDupScoreSQL() . ") AS dupscore" .
+        " FROM patient_data AS p1, patient_data AS p2" .
+        " WHERE p1.dupscore = -9 AND p2.pid < p1.pid" .
+        " GROUP BY p1.pid ORDER BY p1.pid LIMIT $querylimit";
```
to be safe, would escape $querylimit in case in future this became a user input variable. use the escape_limit() function.

sunsetsystems replied:
Done.

bradymiller commented on interface/patient_file/manage_dup_patients.php:
```diff
@@ -0,0 +1,298 @@
+
+require_once("../globals.php");
+require_once("$srcdir/patient.inc");
+require_once("$srcdir/formatting.inc.php");
```
remove the line above (already included in all scripts via composer.json)

sunsetsystems replied:
Done.

bradymiller commented:
Just noted a couple more minor issues. Then ready to come into codebase :)
(thanks for the very cool feature!)

stephenwaite commented:
![southpark_cheering](https://user-images.githubusercontent.com/5618372/124634978-ec6b3480-de54-11eb-8d7c-d3ca8436083f.gif)
